Add blog post: Introducing the Common AI Provider (#1483)
apache/airflow-site: landing-pages/site/content/en/blog/common-ai-provider/index.md

---
title: "Introducing the Common AI Provider: LLM and AI Agent Support for Apache Airflow"
linkTitle: "Introducing the Common AI Provider"
authors:
  - name: "Kaxil Naik"
    github: "kaxil"
    linkedin: "kaxil"
  - name: "Pavan Kumar Gopidesu"
    github: "gopidesupavan"
    linkedin: "pavan-kumar-gopidesu"
description: "The Common AI Provider adds LLM and AI agent operators to Apache Airflow with 6 operators, 5 toolsets, and 20+ model providers in one package."
tags: [Community, Release]

> date: "2026-04-03"

Author: ```suggestion
date: "2026-04-14"
```

The introspection engine builds JSON Schema from method signatures and enriches tool descriptions from docstrings (Sphinx and Google style). You explicitly declare which methods to expose. No auto-discovery, no unintended access. The agent sees `s3_list_keys`, `s3_read_key`, `s3_check_for_key` as typed tools with parameter descriptions pulled straight from the hook.

This works with _any_ hook. Want your agent to send Slack messages? `HookToolset(SlackHook(...), allowed_methods=["send_message"])`. Query Snowflake? Use `SQLToolset` with a Snowflake connection. Hit an internal API? `HookToolset(HttpHook(...), allowed_methods=["run"])`.

You can also compose multiple toolsets in a single agent. Give it `SQLToolset` for database access _and_ `HookToolset` for API calls, and the agent picks the right tool for each step.

Four toolsets ship with the provider:

| Toolset | What it does |
|---|---|
| **`SQLToolset`** | `list_tables`, `get_schema`, `query`, `check_query` for any `DbApiHook` database |
| **`HookToolset`** | Wraps any Airflow hook's methods as agent tools |
| **`MCPToolset`** | Connects to external [MCP](https://modelcontextprotocol.io/) servers via Airflow Connections |

> | **`DataFusionToolset`** | SQL over files in object storage (S3, GCS, Azure Blob) via Apache DataFusion |

Author: ```suggestion
| **`DataFusionToolset`** | SQL over files in object storage (S3, other to come soon) via Apache DataFusion |
```

> | **`DataFusionToolset`** | SQL over files in object storage (S3, GCS, Azure Blob) via Apache DataFusion |

Author: S3 is available though, yeah?

> | **`DataFusionToolset`** | SQL over files in object storage (S3, GCS, Azure Blob) via Apache DataFusion |

Reviewer: Yeah available :) lets keep it
